### PR TITLE
Roles and permissions

### PR DESCRIPTION
**Pull request by kulmann**
## Work In Progress 🚧

Implementation for https://github.com/owncloud/product/issues/85
Also fixes https://github.com/owncloud/ocis-settings/issues/32

Required cleanup before review/merge:
- [x] bring back the [commented tests](https://github.com/owncloud/ocis-settings/blob/roles-and-permissions/pkg/proto/v0/settings.pb.micro_test.go)
- [ ] document folder structure of bundles and values
  - [ ] tracker: https://github.com/owncloud/ocis-settings/issues/13
  - [ ] add to hugo docs
- [x] replace `path` with `filepath`
- [ ] protect all API endpoints with respective permissions
   - [ ] on writing values (especially when the values are meant to be global, i.e. when setting has resource type `SYSTEM`)
   - [x] on listing settings bundles (separate from listing roles)
   - [ ] on saving a bundle
   - [ ] on reading a bundle
   - [ ] on adding a setting to a bundle
   - [ ] on removing a setting from a bundle
   - [ ] on listing values
   - [ ] on saving a value
   - [ ] on reading a value
   - [ ] on listing roles
   - [x] on listing role assignments
   - [ ] on adding role assignments
   - [ ] on removing role assignments
- [ ] additional check: is the authenticated user (account UUID in the access token) allowed to perform the request for the account UUID in the request?
- [ ] avoid duplicates when adding role assignments
  - [ ] we can write a [red test here](https://github.com/owncloud/ocis-settings/blob/roles-and-permissions/pkg/store/filesystem/role_assignments_test.go)

Create and merge pull requests for adapting to the new data model and using the new roles & permissions system in
- [ ] [ocis-accounts](https://github.com/owncloud/ocis-accounts/pull/54): adapt to new keys in settings and settings bundles
- [ ] [ocis-proxy](https://github.com/owncloud/ocis-proxy/pull/70): new middleware for fetching roles and minting them into the token
- [ ] ocis-hello: change registering settings bundles to new `identifier`
- [ ] [owncloud-sdk](https://github.com/owncloud/owncloud-sdk/pull/520): copy the regenerated api client from ocis-settings into the vendor dir and adapt code to new `identifier`
- [ ] [phoenix](https://github.com/owncloud/phoenix/pull/3806): release with new owncloud-sdk and data model changes
- [ ] ocis-phoenix: release with new phoenix

### CODE REVIEW

File: pkg/flagset/flagset.go [hunk 1]
```diff
@@ -159,6 +159,7 @@ func ServerWithConfig(cfg *config.Config) []cli.Flag {
 		},
 		&cli.StringFlag{
 			Name:        "mount-path",
+			Value:       "/var/tmp/ocis-settings",
```
Comment (refs) [inline pkg/flagset/flagset.go:162]: unrelated to this PR, but if we want to make this directories future proof, we must not rely on `/var/tmp/` to exist, and instead use the parent Reva config as Basepath. I'll open an issue for it.

Comment (refs) [reply pkg/flagset/flagset.go:162]: @kulmann when I start the settings service I got two folders that relate to the settings service:

```
drwx------  3 aunger  wheel    96B Jul  8 11:04 ocis-settings
drwx------  3 aunger  wheel    96B Jul  8 11:04 ocis-settings-store
```

Should we align these two? Source is here: https://github.com/owncloud/ocis-settings/blob/master/pkg/store/filesystem/store.go#L15

**Comment by refs** [comment]
@kulmann made some changes to align the mount path for the settings and as a result got the following tree:

```
/var/tmp/ocis-settings
❯ tree .
.
├── bundles
│   └── ocis-settings
│       ├── admin.json
│       ├── guest.json
│       └── user.json
└── ocis-settings
    └── bundles
        └── ocis-settings
            ├── admin.json
            ├── guest.json
            └── user.json

5 directories, 6 files
```

the top level `bundles` seem redundant. Fixing it. Will send you a diff for you to apply.

**Comment by refs** [comment]
With my patch now the folder structure is:


```
/var/tmp/ocis-settings
❯ tree /var/tmp/ocis-settings
/var/tmp/ocis-settings
└── bundles
    └── ocis-settings
        ├── admin.json
        ├── guest.json
        └── user.json

2 directories, 3 files
```

Testing to check the core functionality doesn't break.

File: pkg/flagset/flagset.go (hunk 1 above)
Comment (kulmann) [reply pkg/flagset/flagset.go:162]: As discussed, this happened by accident.... I wanted to align ocis-settings with ocis-accounts, which doesn't have a mount path, but only a data path. Forgot to change how the variable is used in the store implementation. 👀

**Comment by refs** [comment]
bug: add validation when attempting to assign a role that the user already has:

This is a transcript of my session:
```
~/code
❯ micro call com.owncloud.api.settings RoleService.ListRoleAssignments '{"account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51"}'
{
        "assignments": [
                {
                        "id": "721a5cce-a750-47b2-93c2-5decbab1aa87",
                        "account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51",
                        "role_id": "d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11"
                },
                {
                        "id": "9490d020-cfb2-46d8-804c-6b6d088adb62",
                        "account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51",
                        "role_id": "71881883-1768-46bd-a24d-a356a2afdf7f"
                }
        ]
}

~/code
❯ micro call com.owncloud.api.settings RoleService.AssignRoleToUser '{"assignment": {"account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51", "role_id": "d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11"}}'
{
        "assignment": {
                "id": "cf1b10a9-ba30-432e-b3da-4f6be71bcfb1",
                "account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51",
                "role_id": "d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11"
        }
}

~/code
❯ micro call com.owncloud.api.settings RoleService.ListRoleAssignments '{"account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51"}'
{
        "assignments": [
                {
                        "id": "721a5cce-a750-47b2-93c2-5decbab1aa87",
                        "account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51",
                        "role_id": "d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11"
                },
                {
                        "id": "9490d020-cfb2-46d8-804c-6b6d088adb62",
                        "account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51",
                        "role_id": "71881883-1768-46bd-a24d-a356a2afdf7f"
                },
                {
                        "id": "cf1b10a9-ba30-432e-b3da-4f6be71bcfb1",
                        "account_uuid": "4c510ada-c86b-4815-8820-42cdf82c3d51",
                        "role_id": "d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11"
                }
        ]
}
```

**Comment by refs** [comment]
There is no way to prevent assigning a role that is already assigned to the user.

**Comment by refs** [comment]
@kulmann Do we need the account uuid on `ListRoles`? or can we remove it? We can use [google.protobuf.Empty](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/empty)

**Comment by kulmann** [comment]
> There is no way to prevent assigning a role that is already assigned to the user. Note that role with id: `d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11` occurs twice for the same account.

True, this is still work in progress. There is a `// TODO: we need to search for existing role assignments by roleId and accountUuid to avoid duplicate assignments.` in the code. ;-) I will add the TODOs as checkboxes to the cleanup section in the PR description.

**Comment by kulmann** [comment]
> @kulmann Do we need the account uuid on `ListRoles`? or can we remove it? We can use [google.protobuf.Empty](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/empty)

Depends. I guess we should not expose the existing roles to just anyone, but instead only return roles if the authenticated account has a role which is allowed to list roles. I will add a TODO (to the PR and to the description), then we can at least discuss it.

**Comment by kulmann** [comment]
For the record, we moved the code into four separate PRs:
- https://github.com/owncloud/ocis-settings/pull/46
- https://github.com/owncloud/ocis-settings/pull/47
- https://github.com/owncloud/ocis-settings/pull/48
- https://github.com/owncloud/ocis-settings/pull/49

**Comment by kulmann** [comment]
Tracked account uuid validation here https://github.com/owncloud/ocis-settings/issues/50

**Comment by kulmann** [comment]
this entire PR was split into multiple other PRs and is not needed anymore.

**Pull request closed without merging**